### ISSUE TITLE
How to give iam permissions to query created with `@key` directive?

### ISSUE THREAD

**Issue by CaptainChemist** [aws-amplify/amplify-cli#4995]
**Which Category is your question related to?**

AWS Amplify AppSync schema

**Amplify CLI Version** 

4.12.0

**What AWS Services are you utilizing?**

Appsync and DynamoDB

**Provide additional details e.g. code snippets**

I have a model of type `coupon`. I want Users and an iam role to be able to run the `couponByName` query. I add the query using the `@key` directive and the users can run it perfectly. The iam user gets an access denied error though even when I give it permissions using `@auth`.  

```
type Coupon
  @key(name: "nameIndex", fields: ["name"], queryField: "couponByName")
  @model(queries: { get: null })
  @auth(
    rules: [
      { allow: private, provider: iam }
      { allow: groups, groups: ["User"], operations: [read] }
    ]
  ) {
  id: ID!
  createdAt: String
  updatedAt: String
  name: String

}
```

How do you control the permissions of queries created with the `@key` directive so that my iam in my lambda function can run the `couponByName` query? I'm finding it difficult to find examples to find grain permissions on added queries because they don't seem to fit into the "list" "get" types that are described in the documentation. Thanks!

**Comment by CaptainChemist** [comment]
I ended up figuring out a solution. For future people coming across this, I have a few tips when setting permissions:

1) Changing the primary key from `id` to `name` allowed me to directly look up a coupon by name, which is what I was doing before, but critically I could guarantee that I would only ever get back one item, since it was the primary key, rather than a list.

2) Since I knew I was getting back one, not many items back I could safely disable `list` queries for the `User` which prohibits them from listing all the coupons that are available- they need to call one by name specifically. I know that using queries/mutations in a rule is deprecated but there doesn't appear to be a good way to allow reading one but not many coupons otherwise.

3) For some reason there is still this discrepancy between the permissions granted to my default cognito users versus iam users but that only shows up using these secondary keys. By eliminating the need for a secondary key query altogether and using the name as the primary key, the permissions of the iam matched what I set it to be in the `@auth` directive. 

This was the updated model that worked well for me. Hopefully, it helps out someone else.

```
type Coupon
  @key(fields: ["name"])
  @model
  @auth(
    rules: [
      { allow: private, provider: iam }
      { allow: groups, groups: ["User"], queries: [get], mutations: null }
    ]
  ) {
  createdAt: String
  updatedAt: String
  name: String!
}
```
